src/arch/ia32/descriptor/gate/trap.rs

```diff
@@ -1,41 +1,91 @@
-use super::lower::Lower;
+//! A module containing the implementation and the tests for the [`TrapGate`]
+//! structure along some utility types revolving around those.
 use super::{GateSize, PrivilegeLevel};
 use crate::arch::ia32::selector::SegmentSelector;
-use upper::Upper;
+use bit_field::BitField;
+use configuration::Configuration;
 
-mod upper;
+mod configuration;
 
+/// Trap gate structure that can be converted into a generic gate structure
+/// once set up.
 #[must_use]
 #[derive(Default, Copy, Clone)]
 #[repr(C, packed)]
 pub struct TrapGate {
-    pub lower: Lower,
-    pub upper: Upper,
+    /// Bits 0 to 15 of the trap routine address.
+    offset_15_0: u16,
+    /// The segment selector pointing to the segment to use during the trap
+    /// routine.
+    segment_selector: SegmentSelector,
+    /// The trap descriptor configuration.
+    configuration: Configuration,
+    /// Bits 16 to 31 of the trap routine address.
+    offset_31_16: u16,
 }
 
 impl TrapGate {
+    /// Creates a new [`TrapGate`] structure from a given trap routine
+    /// address and segment selector.
+    ///
+    /// # Arguments
+    ///
+    /// * `segment_selector` - The segment selector used by the gate.
+    ///
+    /// # Note
+    ///
+    /// The present bit will be enabled when using this constructor.
     pub fn new(offset: u32, segment_selector: SegmentSelector) -> Self {
-        let offset_15_0 = offset & 0xFFFF;
-        let offset_31_16 = offset >> 16;
         TrapGate {
-            lower: Lower::default()
-                .offset_low(offset_15_0)
-                .segment_selector(segment_selector.into()),
-            upper: Upper::default().offset_high(offset_31_16).present(1),
+            offset_15_0: offset.get_bits(0..16).try_into().unwrap(),
+            offset_31_16: offset.get_bits(16..32).try_into().unwrap(),
+            configuration: Configuration::default().present(true),
+            segment_selector,
         }
     }
 
+    /// Change a [``]'s privilege level by another one.
```

Comment:
It think the `TrapGate` word is missing between the brackets.

Reply:
Done in 801000d